protected/humhub/services/MigrationService.php

```diff
@@ -0,0 +1,328 @@
+use humhub\components\Event;
+use humhub\components\Module;
+use humhub\events\MigrationEvent;
+use humhub\interfaces\ApplicationInterface;
+use Throwable;
+use Yii;
+use yii\base\ActionEvent;
+use yii\base\Component;
+use yii\base\Controller;
+use yii\base\InvalidConfigException;
+use yii\base\Module as BaseModule;
+use yii\console\ExitCode;
+
+/**
+ * @since 1.16
+ */
+class MigrationService extends Component
+{
+    public const EVENT_AFTER_MIGRATION = 'afterMigration';
+    public const EVENT_BEFORE_MIGRATION = 'beforeMigration';
+    protected const MIGRATION_NEW = 'new';
+    protected const MIGRATION_UNINSTALL = 'uninstall';
+    protected const MIGRATION_UP = 'up';
+
+    protected BaseModule $module;
+    protected ?string $path = null;
+    private ?int $lastMigrationResult = null;
+    private ?string $lastMigrationOutput = null;
+
+    /**
+     * @param Module|ApplicationInterface|Application|null $module
+     * @param array $config
+     */
+    public function __construct(?BaseModule $module = null, array $config = [])
+    {
+        $this->module = $module ?? Yii::$app;
+        parent::__construct($config);
+    }
+
+    public function init()
+    {
+        parent::init();
+
+        if ($this->path === null) {
+            $moduleId = $this->module instanceof Module
+                ? $this->module->id
+                : 'humhub';
+
+            $this->path = "@$moduleId/migrations";
+        }
+
+        $realpath = $this->getPath(true);
+
+        if ($realpath === false || !is_dir($realpath)) {
+            Yii::debug("Module has no migrations directory.", $this->module->id);
+            $this->path = null;
+        }
+    }
+
+    /**
+     * @return Module|ApplicationInterface|Application|null
+     */
+    public function getModule(): BaseModule
+    {
+        return $this->module;
+    }
+
+    public function getPath(bool $resolve = false): ?string
+    {
+        if (!$resolve || $this->path === null) {
+            return $this->path;
+        }
+
+        $path = realpath(Yii::getAlias($this->path));
+
+        if ($path === false) {
+            return null;
+        }
+
+        return $path;
+    }
+
+    public function getLastMigrationOutput(): ?string
+    {
+        return $this->lastMigrationOutput;
+    }
+
+    public function getLastMigrationResult(): ?int
+    {
+        return $this->lastMigrationResult;
+    }
+
+    public function isValid(): bool
+    {
+        return $this->path !== null;
+    }
+
+    /**
+     * Run migrations.
+     *
+     * @param string|null $output Output of the migration, if available
+     */
+    public function migrateNew(?string &$output = null): ?bool
+    {
+        return $this->migrate(MigrationService::MIGRATION_NEW, $output);
+    }
+
+    /**
+     * Check for pending migrations
+     *
+     * @param string|null $output Output of the migration, if available
+     */
+    public function migrateUp(?string &$output = null): ?bool
```

Comment:
Why should the output of a migration be passed to the migrate command? Same for `migrateNew` and following methods.

Reply:
For convenience. Since it's just a reference, it does not consume any (considerable amount of) memory.

But to have the output of the migration makes the call in the `InformationController` cheaper:
```php
MigrationService::create()->migrateUp($migrationOutput);
```
rather than:
```php
$service = MigrationService::create();
$service->migrateUp();
$migrationOutput = $service->getLastMigrationOutput()
```

Reply:
Ok, I see. Thank you.

I still prefer the second option. This makes the code easier to understand and we don't need more in-depth optimizations here.